addons/l10n_vn_edi/models/account_move.py

```diff
@@ -0,0 +1,719 @@
+        """ Send a request to cancel the invoice. """
+        self.ensure_one()
+        _request_response, error_message = self._l10n_vn_edi_send_request(
+            method='POST',
+            endpoint='InvoiceAPI/InvoiceWS/cancelTransactionInvoice',
+            params={
+                'supplierTaxCode': self.company_id.vat,
+                'templateCode': self.l10n_vn_edi_invoice_symbol.invoice_template_id.name,
+                'invoiceNo': self.l10n_vn_edi_invoice_number,
+                'strIssueDate': self._l10n_vn_edi_format_date(self.l10n_vn_edi_issue_date),
+                'additionalReferenceDesc': agreement_document_name,
+                'additionalReferenceDate': self._l10n_vn_edi_format_date(agreement_document_date),
+                'reasonDelete': reason,
+            },
+        )
+        if error_message:
+            raise UserError(error_message)
+
+        self.l10n_vn_edi_state = 'cancelled'
+        self.with_context(no_new_invoice=True).message_post(
+            body=_('The invoice has been cancelled for reason: %s', reason),
+        )
+        self.button_cancel()
+
+    def button_draft(self):
+        # EXTEND account
+        # When going from cancelled => draft, we ensure to clear the edi fields so that the invoice can be resent if required.
+        sinvoices = self.filtered(
+            lambda i: i.company_id.country_id.code == 'VN' and i.state == 'cancel'
+        )
+        sinvoices.write({
+            'l10n_vn_edi_transaction_id': False,
+            'l10n_vn_edi_invoice_number': False,
+            'l10n_vn_edi_reservation_code': False,
+            'l10n_vn_edi_issue_date': False,
+            'l10n_vn_edi_state': False,
+        })
+        # Cleanup the files as well. They will still be available in the chatter.
+        sinvoices.l10n_vn_edi_sinvoice_xml_file_id.unlink()
+        sinvoices.l10n_vn_edi_sinvoice_pdf_file_id.unlink()
+        sinvoices.l10n_vn_edi_sinvoice_file_id.unlink()
+        return super().button_draft()
+
+    def _l10n_vn_edi_generate_invoice_json(self):
+        """ Return the dict of data that will be sent to the api in order to create the invoice. """
+        # We leave the summarized information computation to SInvoice.
+        self.ensure_one()
+        # This MUST match chronologically with the sequence they generate on their system, which is why it is set to now.
+        self.l10n_vn_edi_issue_date = fields.Datetime.now()
+        return {
+            'generalInvoiceInfo': self._l10n_vn_edi_get_general_invoice_information(),
+            'buyerInfo': self._l10n_vn_edi_get_buyer_information(),
+            'sellerInfo': self._l10n_vn_edi_get_seller_information(),
+            'payments': self._l10n_vn_edi_get_payment_information(),
+            'itemInfo': self._l10n_vn_edi_get_item_information(),
+            'taxBreakdowns': self._l10n_vn_edi_get_tax_breakdowns(),
+        }
```

Comment:
This is usually a bad idea because in case of changes in the json/fix etc, sometimes, you could need a value in a sub method in another one. What about a more constructive approach:
```py
json_values = {}
self._l10n_vn_edi_add_general_invoice_information(json_values)
...
```